CorsixTH/Src/th_movie.cpp

```diff
@@ -296,59 +247,28 @@ int movie_picture_buffer::write(AVFrame* pFrame, double dPts) {
   return 0;
 }
 
-av_packet_queue::av_packet_queue()
-    : first_packet(nullptr), last_packet(nullptr), count(0), mutex{}, cond{} {}
-
-int av_packet_queue::get_count() const { return count; }
+av_packet_queue::av_packet_queue() : data{}, mutex{}, cond{} {}
 
-void av_packet_queue::push(AVPacket* pPacket) {
-#if (defined(CORSIX_TH_USE_LIBAV) &&                       \
-     LIBAVCODEC_VERSION_INT < AV_VERSION_INT(57, 8, 0)) || \
-    (defined(CORSIX_TH_USE_FFMPEG) &&                      \
-     LIBAVCODEC_VERSION_INT < AV_VERSION_INT(57, 12, 100))
-  if (av_dup_packet(pPacket) < 0) {
-    throw std::runtime_error("Failed to duplicate AV packet");
-  }
-#endif
-
-  th_packet_list* pNode = (th_packet_list*)av_malloc(sizeof(th_packet_list));
-  pNode->pkt = *pPacket;
-  pNode->next = nullptr;
+std::size_t av_packet_queue::get_count() const { return data.size(); }
 
+void av_packet_queue::push(av_packet_unique_ptr pPacket) {
   std::lock_guard<std::mutex> lock(mutex);
-
-  if (last_packet == nullptr) {
-    first_packet = pNode;
-  } else {
-    last_packet->next = pNode;
-  }
-  last_packet = pNode;
-  count++;
+  data.push(std::move(pPacket));
 
   cond.notify_one();
 }
 
-AVPacket* av_packet_queue::pull(bool fBlock) {
+av_packet_unique_ptr av_packet_queue::pull(bool block) {
   std::unique_lock<std::mutex> lock(mutex);
 
-  th_packet_list* pNode = first_packet;
-  if (pNode == nullptr && fBlock) {
+  if (data.empty() && block) {
     cond.wait(lock);
-    pNode = first_packet;
   }
 
-  AVPacket* pPacket;
-  if (pNode == nullptr) {
-    pPacket = nullptr;
-  } else {
-    first_packet = pNode->next;
-    if (first_packet == nullptr) {
-      last_packet = nullptr;
-    }
-    count--;
-    pPacket = (AVPacket*)av_malloc(sizeof(AVPacket));
-    *pPacket = pNode->pkt;
-    av_free(pNode);
+  av_packet_unique_ptr pPacket(nullptr);
+  if (!data.empty()) {
+    pPacket.swap(data.front());
+    data.pop();
   }
 
   return pPacket;
```

Comment:
Old problem, but this looks weird. It locks the mutex at the start of the function but doesn't release it here.

Reply:
Yes it does. C++ mutexes release at end of scope.